### PR TITLE
JIT Cleanups

### PR DESCRIPTION
**Pull request by zdevito**
1. Change scope ownership model so they can be shared across Graphs.
   Now scopes own their parent and are intrusive pointers. Graphs
   no longer require a scope_root and cloning a node automatically
   clones its scope. This causes some changes in expect files for
   trace+script things. As far as I can tell these are not bugs but
   a different way of interpreting how scopes should propagate.
   Big traces like that of alexnet keep their scopes unchanged.
2. Remove VariableType.cpp dependency on a symbol being in the pre-
   declared symbol list.

### CODE REVIEW

**Comment by SsnL** [comment]
Thank you!

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -135,7 +135,8 @@
 std::shared_ptr<jit::tracer::TracingState> tracer_state;
 if (jit::tracer::isTracing()) {
   tracer_state = jit::tracer::getTracingState();
-  node = tracer_state->graph->create(jit::aten::${trace_name}, /*num_outputs=*/0);
+  const auto op_name = jit::Symbol::fromQualString("aten::${trace_name}");
```
Comment (dzhulgakov) [inline tools/autograd/gen_variable_type.py]: Making it static might or might not make things a bit fastee

File: aten/src/ATen/core/intrusive_ptr.h [hunk 2]
```diff
@@ -138,9 +138,12 @@ template <
     class NullType = detail::intrusive_target_default_null_type<TTarget>>
 class C10_EXPORT intrusive_ptr final {
  private:
-  static_assert(
-      std::is_base_of<intrusive_ptr_target, TTarget>::value,
-      "intrusive_ptr can only be used for classes that inherit from intrusive_ptr_target.");
+//  the following static assert would be nice to have but it requires
+//  the target class T to be fully defined when intrusive_ptr<T> is instanciated
```
Comment (orionr) [inline aten/src/ATen/core/intrusive_ptr.h]: Nit: instantiated.

File: torch/csrc/jit/ir.h [hunk 3]
```diff
@@ -105,39 +105,48 @@ struct Use {
 // The trie never needs to shrink, it only grows until it is disposed
 // of when Graph is deallocated. Hence, pointers to scopes held by nodes
 // will always be valid as long as Graph is alive.
-struct Scope {
+struct Scope;
+using ScopePtr = c10::intrusive_ptr<Scope>;
+
+struct TORCH_API Scope : public c10::intrusive_ptr_target {
```
Comment (orionr) [inline torch/csrc/jit/ir.h]: W00t.

File: torch/csrc/jit/ir.h [hunk 4]
```diff
@@ -105,39 +105,48 @@ struct Use {
...
 private:
-  Scope* parent_;
+  ScopePtr parent_;
   Symbol name_;
-  std::vector<std::unique_ptr<Scope> > children_;
+  ScopePtr intrusive_from_this() {
+    c10::raw::intrusive_ptr::incref(this); // :(
```
Comment (orionr) [inline torch/csrc/jit/ir.h]: Do you want to put a bit more comment on this one? :)

File: torch/csrc/jit/passes/canonicalize.cpp [hunk 5]
```diff
@@ -5,7 +5,7 @@ namespace torch { namespace jit {
 // Canonicalize a graph, renumbering it so that all structurally equivalent
 // graphs have same numbers.
 std::shared_ptr<Graph> Canonicalize(const std::shared_ptr<Graph>& graph) {
-  auto r = std::make_shared<Graph>(graph->scope_root());
+  auto r = std::make_shared<Graph>(graph->current_scope());
```
Comment (orionr) [inline torch/csrc/jit/passes/canonicalize.cpp]: This seems like a logical change?

File: torch/csrc/jit/passes/onnx.cpp [hunk 6]
```diff
@@ -12,7 +12,7 @@ namespace torch { namespace jit {
 
 // Transform PythonOps into Nodes that match ONNX semantics.
 std::shared_ptr<Graph> ToONNX(std::shared_ptr<Graph>& graph, ::torch::onnx::OperatorExportTypes operator_export_type) {
-  auto new_graph = std::make_shared<Graph>(graph->scope_root());
+  auto new_graph = std::make_shared<Graph>(graph->current_scope());
```
Comment (orionr) [inline torch/csrc/jit/passes/onnx.cpp]: Same.

**Review by orionr — APPROVED** [review]
Nice!

**Review by apaszke — APPROVED** [review]

File: torch/csrc/jit/passes/canonicalize.cpp (hunk 5 above)
Comment (zdevito) [reply torch/csrc/jit/passes/canonicalize.cpp]: I think so -- basically the only previous reason to copy the scope root was to allow for scopes to be copied over. This is probably unnecessary now but it ensures that the current_scope is not set differently across two graph copies, which seems reasonable.

File: tools/autograd/gen_variable_type.py (hunk 1 above)
Comment (zdevito) [reply tools/autograd/gen_variable_type.py]: Yeah, I though I did write static, but I must have deleted it.

**Pull request closed without merging**